Document new features for Buildkite hosted agents GA release (#2786)
buildkite/docs: pages/pipelines/buildkite_secrets.md

- Are available to both [Buildkite hosted](/docs/pipelines/hosted-agents/overview) as well as self-hosted agents.

## Create a secret

### Using the Buildkite interface

To create a new Buildkite secret using the Buildkite interface:

1. Select **Agents** in the global navigation to access the **Clusters** page.
1. Select the cluster in which to create the new Buildkite secret.
1. Select **Secrets** to access the **Secrets** page, then select **New Secret**.
1. Enter a **Key** for the secret, which can only contain letters, numbers, and underscores, as valid characters.

    **Notes:**
    * The maximum permitted length for a key is 255 characters.
    * If you attempt to use any other characters for the key, or you begin your key with `buildkite` or `bk` (regardless of case), your secret will not be created when selecting **Create Secret**.

1. Enter the **Value** for the secret. This value can be any number of valid UTF-8 characters up to a maximum of 8 kilobytes. Be aware that once the secret is created, this value will no longer be visible through the Buildkite interface and will be redacted when output in build logs.

> 1. Select **Create Secret** to create your new secret, which can now be referenced and used in build steps and jobs, respectively.

Reviewer: ```suggestion
1. Select **Create Secret** to create your new secret, which can now be accessed within jobs through the `buildkite-agent secret get` command.
```

    `SECRET_VAR=$(buildkite-agent secret get secret_name)`

- Redirecting the value of a Buildkite secret with the key `secret_name` into a file called `secret.txt`:

    `buildkite-agent secret get secret_name > secret.txt`

- Passing the output of your Buildkite secret (via the `buildkite-agent secret get` command) to your own tool named `cli-tool` that accepts a secret via its `-token` option:

    `cli-tool —token $(buildkite-agent secret get secret_name)`

Here’s a simple example of how one of these commands might appear in a Buildkite Pipeline step:

```yaml
steps:
  - agents: { queue: "hosted" }
    command:
      - buildkite-agent secret get secret_name > secret.txt
```
{: codeblock-file="pipeline.yml"}

## Redaction

If any Pipeline, script or your own tool (accidentally) prints out the value of any Buildkite secret to standard out, this value is automatically redacted from the build logs.

If for any reason you detect a secret value that isn't redacted, please rotate your secrets and contact security@buildkite.com.

## Security

Buildkite secrets are designed, with the following controls in place:

> - Secrets are encrypted in transit using TLS.
> - Secrets are always stored encrypted at rest.

Reviewer 2: i think we should be as specific as possible with crypto stuff, otherwise it might come across as "don't worry, your secrets are encrypted! trust us! ;)"

maybe something like:
```suggestion
- Secrets are always stored encrypted at rest using [AES-256 GCM](https://en.wikipedia.org/wiki/Galois/Counter_Mode) encryption. <!-- maybe link to something more authoritative than wikipedia? idk -->
- Encryption keys are managed with AWS KMS, and are scoped per-cluster. This means that were an encryption key to leak, blast radius is limited to only the secrets for that cluster.
- Secrets are encrypted in transit using TLS.
```

> - Secrets are always stored encrypted at rest.

Author: Thanks @moskyb ! Happy to add this.

Out of interest, by "blast radius", you mean the scope of the leak?

> - Secrets are always stored encrypted at rest.

Reviewer 2: yeah - if an encryption token for a given cluster gets leaked, that token can only be used to decrypt secrets in that cluster - other secrets couldn't be decrypted by that token

> - Secrets are always stored encrypted at rest.

Reviewer: I don't think we should include this level of detail at the moment. We can revisit including more detail in the future.

- All access to the secrets are logged.
- Employee access to secrets is strictly limited and audited.

> 📘 Buildkite secrets is not a zero-knowledge system

> > Buildkite owns, stores and manages the keys used for encrypting the secrets stored in the service at rest and in transit. Buildkite recommend customers implement additional controls to manage the lifecycle of secrets stored within Buildkite secrets, in addition to any monitoring capability you may require in line with your risk appetite.

Reviewer 2: we give customers the necessary information here, but i think we should draw the conclusion as well. absolute candor, and all that

```suggestion
> Buildkite owns, stores and manages the keys used for encrypting the secrets stored in the service at rest and in transit. **This means that were Buildkite to be compromised, your secrets could be compromised as well.** Buildkite recommend customers implement additional controls to manage the lifecycle of secrets stored within Buildkite secrets, in addition to any monitoring capability you may require in line with your risk appetite.
```

## Best practices

> - Do not store long lived secrets or secrets with no expiry in Buildkite secrets.
> - Only store short lived secrets in Buildkite secrets.

Reviewer 2: i think these two points are basically saying the same thing, could probably have just the first:
```suggestion
- Do not store long lived secrets or secrets with no expiry in Buildkite secrets.
```

Author: Yeah, good point, I'll combine these two.

- Track the secrets stored in Buildkite secrets within your own asset management processes.
- Enable logging for services that are accessed using the secrets stored in Buildkite secrets.
- Should you detect a compromise or are concerned about the security of your secrets, please contact security@buildkite.com immediately.

## Alternative secret management methods

> If Buildkite secrets is not suitable for your requirements, Buildkite recommends that you manage your own secrets in your own infrastructure, and then implement [OIDC](/docs/pipelines/security/oidc) with your cloud storage/service provider to grant hosted agents access to these secrets. You can follow the [OIDC with AWS](/docs/pipelines/security/oidc/aws) process to work with secrets in AWS managed by the AWS Systems Manager (SSM). These secrets can then be accessed in your Buildkite Pipelines using hosted agents.

Reviewer 2: ```suggestion
If Buildkite secrets is not suitable for your requirements, Buildkite recommends that you manage your own secrets in your own infrastructure, and then implement [OIDC](/docs/pipelines/security/oidc) with your cloud storage/service provider to grant hosted agents access to these secrets. You can follow the [OIDC with AWS](/docs/pipelines/security/oidc/aws) process to work with secrets in AWS managed by the AWS Systems Manager (SSM). These secrets can then be accessed in your Buildkite Pipelines using both hosted and self-hosted agents.
```

Author: @moskyb , could we also change:

> to grant hosted agents access to these secrets

to:

> to grant hosted and self-hosted agents access to these secrets

?
